### PR TITLE
Cache - first version of infrastructure and SearchStats implementation

### PR DESCRIPTION
**Pull request by edoshor**

### CODE REVIEW

**Comment by bbfsdev** [comment]
Nice, looks good to me. You cam merge.

File: cache/search_stats.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/volatiletech/sqlboiler/queries"
+	"gopkg.in/volatiletech/null.v6"
+
+	"github.com/Bnei-Baruch/archive-backend/mdb"
+)
+
+type Histogram map[string]int
+
+func (h Histogram) Increment(k string, v int) {
+	h[k] += v
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/search_stats.go]: Not needed as we get the default value (which is 0 for int) if the key is not in the map.
Take a look at [this post from the official go blog](https://blog.golang.org/go-maps-in-action)

File: cache/search_stats.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		h.Increment(k, v)
+	}
+}
+
+type StatsNode struct {
+	id       int64
+	parentID int64
+	uid      string
+	children []int64
+	hist     Histogram
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/search_stats.go]: This is a struct for a single node in an hierarchical classification tree, i.e. tags / sources.
MDB employ a UNIQUE constraint at the DB level.
As long as developers won't mix those in the same `StatsTree` I don't expect problems.

File: cache/core.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	cm.refreshIntervals = make(map[string]int64, len(refreshIntervals))
+	for k, v := range refreshIntervals {
+		cm.refreshIntervals[k] = int64(v.Truncate(time.Second).Seconds())
+	}
+
+	cm.ticker = time.NewTicker(time.Second)
+	go func() {
+		for range cm.ticker.C {
+			cm.ticks++
+			if cm.ticks%cm.refreshIntervals["SearchStats"] == 0 {
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/core.go]: go fmt. not me

File: cache/core.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+				cm.refresh(cm.ssc)
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/core.go]: We should if we have something useful to do with them.
Currently, I'm just logging it in `CacheManagerImpl.refresh`
We should log these to rollbar one day

File: cache/core.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	cm.ticker.Stop()
+}
+
+func (cm *CacheManagerImpl) SearchStats() SearchStatsCache {
+	return cm.ssc
+}
+
+func (cm *CacheManagerImpl) refresh(p Provider) {
+	log.Infof("Refreshing %s", p)
+	if err := p.Refresh(); err != nil {
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/core.go]: You're correct we should log these to rollbar

File: cache/search_stats.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+		} else if k[0] == 's' {
+			tmp = sources
+		}
+
+		var ctName = ""
+		if typeID.Valid {
+			ct, ok := mdb.CONTENT_TYPE_REGISTRY.ByID[typeID.Int64]
+			if ok {
+				ctName = ct.Name
+			} else {
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/search_stats.go]: Though I don't expect such cases I want to think it's ok to ignore such cases silently in this context.

File: cache/search_stats.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	byID map[int64]*StatsNode
+}
+
+func NewStatsTree() *StatsTree {
+	st := new(StatsTree)
+	st.byID = make(map[int64]*StatsNode)
+	return st
+}
+
+func (st *StatsTree) accumulate() {
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/search_stats.go]: will do

File: cache/search_stats.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+		s = make([]int64, len(parents))
+		i := 0
+		for k := range parents {
+			s[i] = k
+			i++
+		}
+	}
+}
+
+func (st *StatsTree) flatten() map[string]Histogram {
```
[the remark this answers is not in the document]
Comment (edoshor) [reply cache/search_stats.go]: will do

**Pull request merged**